#### This is an EDA that seeks to answer questions regarding patterns in a dataset of cnv probes on cancer associated genes
#### It ends in a suggestion to weight multiple cnvs and calls to get a useful concensus call.

#### For each gene, what proportion of patients have probe regions that are called loss, neutral, or gain?

In [1]:
import pandas as pd #data procesing

In [2]:
data = pd.read_csv(r"C:\Users\raobr\Documents\Tempus\Junior Variant Scientist\CNV_challenge (5) Brian Rao.csv")

In [3]:
data.head()

orderid  gene  probe  cnv     call
0    PAT0  TP53    1.0  2.0  neutral
1    PAT0  TP53    2.0  2.0  neutral
2    PAT0  TP53    3.0  2.0  neutral
3    PAT0  TP53    4.0  2.0  neutral
4    PAT0  TP53    5.0  2.0  neutral

In [4]:
print(data.describe()) #Confirms what we were told about the data we would find on the probe regions and copy numbers
print(data['orderid'].nunique()) #Tells us how many patients are in the data set

              probe           cnv
count  63086.000000  63086.000000
mean      10.161494      2.840725
std        6.983322      3.092748
min        1.000000      0.000000
25%        4.000000      2.000000
50%        9.000000      2.000000
75%       15.000000      3.000000
max       25.000000     20.000000
1527


In [5]:
df_TP53 = data.loc[data.gene == 'TP53'] #Splits data by each of the 3 genes, TP53,EGFR,CDKN2A
df_EGFR = data.loc[data.gene == 'EGFR']
df_CDKN2A = data.loc[data.gene == 'CDKN2A']


In [6]:
df_TP53  #This preview only shows neutral calls but the next steps will give us a count of the other calls as well.

orderid  gene  probe  cnv     call
0         PAT0  TP53    1.0  2.0  neutral
1         PAT0  TP53    2.0  2.0  neutral
2         PAT0  TP53    3.0  2.0  neutral
3         PAT0  TP53    4.0  2.0  neutral
4         PAT0  TP53    5.0  2.0  neutral
5         PAT0  TP53    6.0  2.0  neutral
6         PAT0  TP53    7.0  2.0  neutral
7         PAT0  TP53    8.0  2.0  neutral
8         PAT0  TP53    9.0  2.0  neutral
9         PAT0  TP53   10.0  2.0  neutral
10        PAT0  TP53   11.0  2.0  neutral
11        PAT0  TP53   12.0  2.0  neutral
41        PAT1  TP53    1.0  6.0  neutral
42        PAT1  TP53    2.0  6.0  neutral
43        PAT1  TP53    3.0  6.0  neutral
44        PAT1  TP53    4.0  6.0  neutral
45        PAT1  TP53    5.0  6.0  neutral
46        PAT1  TP53    6.0  6.0  neutral
47        PAT1  TP53    7.0  6.0  neutral
48        PAT1  TP53    8.0  6.0  neutral
49        PAT1  TP53    9.0  6.0  neutral
50        PAT1  TP53   10.0  6.0  neutral
51        PAT1  TP53   11.0  6.0  neutral
52        PAT1  TP53   12.0  6.0  neutral
53        PAT1  TP53   13.0  6.0  neutral
54        PAT1  TP53   14.0  6.0  neutral
55        PAT1  TP53   15.0  6.0  neutral
85        PAT2  TP53    1.0  2.0  neutral
86        PAT2  TP53    2.0  2.0  neutral
87        PAT2  TP53    3.0  2.0  neutral
...        ...   ...    ...  ...      ...
62969  PAT1524  TP53    8.0  1.0  neutral
62970  PAT1524  TP53    9.0  1.0  neutral
62971  PAT1524  TP53   10.0  1.0  neutral
63001  PAT1525  TP53    1.0  2.0  neutral
63002  PAT1525  TP53    2.0  2.0  neutral
63003  PAT1525  TP53    3.0  2.0  neutral
63004  PAT1525  TP53    4.0  2.0  neutral
63005  PAT1525  TP53    5.0  2.0  neutral
63006  PAT1525  TP53    6.0  2.0  neutral
63007  PAT1525  TP53    7.0  2.0  neutral
63008  PAT1525  TP53    8.0  2.0  neutral
63009  PAT1525  TP53    9.0  2.0  neutral
63010  PAT1525  TP53   10.0  2.0  neutral
63011  PAT1525  TP53   11.0  2.0  neutral
63012  PAT1525  TP53   12.0  2.0  neutral
63042  PAT1526  TP53    1.0  2.0  neutral
63043  PAT1526  TP53    2.0  2.0  neutral
63044  PAT1526  TP53    3.0  2.0  neutral
63045  PAT1526  TP53    4.0  2.0  neutral
63046  PAT1526  TP53    5.0  2.0  neutral
63047  PAT1526  TP53    6.0  2.0  neutral
63048  PAT1526  TP53    7.0  2.0  neutral
63049  PAT1526  TP53    8.0  2.0  neutral
63050  PAT1526  TP53    9.0  2.0  neutral
63051  PAT1526  TP53   10.0  2.0  neutral
63052  PAT1526  TP53   11.0  2.0  neutral
63053  PAT1526  TP53   12.0  2.0  neutral
63054  PAT1526  TP53   13.0  2.0  neutral
63055  PAT1526  TP53   14.0  2.0  neutral
63056  PAT1526  TP53   15.0  2.0  neutral

[20214 rows x 5 columns]

We know 1527 patients are in the dataset total, but we want to confirm how many participated in each gene's tests

In [7]:
df_TP53.nunique() 

orderid    1525
gene          1
probe        15
cnv          12
call          3
dtype: int64

In [8]:
df_EGFR.nunique()

orderid    1472
gene          1
probe        25
cnv          16
call          3
dtype: int64

In [9]:
df_CDKN2A.nunique()

orderid    1522
gene          1
probe         4
cnv          12
call          3
dtype: int64

We are now in a position to answer Question 1. While looking at this one gene, TP53,
dropping all but one call of duplicate types by each patient allows us to get a total count of each call
to calculate the proportion of calls for this gene.

In [10]:
TP53_call_search = df_TP53.drop_duplicates(['orderid','call'], keep='first')
TP53_call_search

orderid  gene  probe  cnv     call
0         PAT0  TP53    1.0  2.0  neutral
41        PAT1  TP53    1.0  6.0  neutral
85        PAT2  TP53    1.0  2.0  neutral
129       PAT3  TP53    1.0  2.0  neutral
168       PAT4  TP53    1.0  1.0  neutral
207       PAT5  TP53    1.0  2.0  neutral
251       PAT6  TP53    1.0  2.0  neutral
263       PAT7  TP53    1.0  3.0  neutral
307       PAT8  TP53    1.0  2.0  neutral
351       PAT9  TP53    1.0  2.0  neutral
395      PAT10  TP53    1.0  5.0  neutral
439      PAT11  TP53    1.0  3.0  neutral
483      PAT12  TP53    1.0  2.0  neutral
527      PAT13  TP53    1.0  3.0  neutral
568      PAT14  TP53    1.0  2.0  neutral
612      PAT15  TP53    1.0  2.0  neutral
656      PAT16  TP53    1.0  4.0  neutral
700      PAT17  TP53    1.0  4.0  neutral
744      PAT18  TP53    1.0  2.0  neutral
788      PAT19  TP53    1.0  2.0  neutral
832      PAT20  TP53    1.0  2.0  neutral
876      PAT21  TP53    1.0  2.0  neutral
920      PAT22  TP53    1.0  2.0  neutral
964      PAT23  TP53    1.0  2.0  neutral
1005     PAT24  TP53    1.0  2.0  neutral
1046     PAT25  TP53    1.0  2.0  neutral
1090     PAT26  TP53    1.0  2.0  neutral
1134     PAT27  TP53    1.0  2.0  neutral
1178     PAT28  TP53    1.0  2.0  neutral
1217     PAT29  TP53    1.0  2.0  neutral
...        ...   ...    ...  ...      ...
61821  PAT1497  TP53    1.0  2.0  neutral
61865  PAT1498  TP53    1.0  1.0  neutral
61909  PAT1499  TP53    1.0  2.0  neutral
61950  PAT1500  TP53    1.0  1.0  neutral
61989  PAT1501  TP53    1.0  2.0  neutral
62033  PAT1502  TP53    1.0  2.0  neutral
62072  PAT1503  TP53    1.0  3.0  neutral
62116  PAT1504  TP53    1.0  2.0  neutral
62160  PAT1505  TP53    1.0  2.0  neutral
62199  PAT1506  TP53    1.0  3.0  neutral
62243  PAT1507  TP53    1.0  2.0  neutral
62287  PAT1508  TP53    1.0  2.0  neutral
62331  PAT1509  TP53    1.0  2.0  neutral
62372  PAT1510  TP53    1.0  2.0  neutral
62416  PAT1511  TP53    1.0  3.0  neutral
62460  PAT1512  TP53    1.0  1.0  neutral
62504  PAT1513  TP53    1.0  3.0  neutral
62545  PAT1514  TP53    1.0  2.0  neutral
62589  PAT1515  TP53    1.0  2.0  neutral
62630  PAT1516  TP53    1.0  2.0  neutral
62671  PAT1517  TP53    1.0  4.0  neutral
62715  PAT1518  TP53    1.0  1.0  neutral
62754  PAT1519  TP53    1.0  2.0  neutral
62795  PAT1520  TP53    1.0  1.0  neutral
62839  PAT1521  TP53    1.0  1.0  neutral
62880  PAT1522  TP53    1.0  2.0  neutral
62921  PAT1523  TP53    1.0  2.0  neutral
62962  PAT1524  TP53    1.0  1.0  neutral
63001  PAT1525  TP53    1.0  2.0  neutral
63042  PAT1526  TP53    1.0  2.0  neutral

[1540 rows x 5 columns]

1540 rows confirms that some patients had more than one type of call

In [11]:
TP53_call_search['call'].value_counts() 

neutral    1502
loss         26
gain         12
Name: call, dtype: int64

Now we just run the same code for other two genes

In [12]:
EGFR_call_search = df_EGFR.drop_duplicates(['orderid','call'], keep='first')
EGFR_call_search['call'].value_counts() 


neutral    1356
gain         85
loss         50
Name: call, dtype: int64

In [13]:
CDKN2A_call_search = df_CDKN2A.drop_duplicates(['orderid','call'], keep='first')
CDKN2A_call_search['call'].value_counts() 


neutral    1337
loss        186
gain          7
Name: call, dtype: int64

#### How many cases have multiple calls per gene?
#### Are there any that have more than one gene with a conflicting call?


We need to search for patients that had more than one call for a given gene, which means they appeared twice on the
list of unique calls we created.

In [14]:
conflicts_TP53 = TP53_call_search[TP53_call_search.duplicated('orderid', keep = False)]
conflicts_TP53

orderid  gene  probe   cnv     call
5229    PAT127  TP53    1.0   0.0     loss
5243    PAT127  TP53   15.0   2.0  neutral
12823   PAT312  TP53    1.0   4.0  neutral
12834   PAT312  TP53   12.0   0.0     loss
18835   PAT458  TP53    1.0   8.0     gain
18836   PAT458  TP53    2.0   3.0  neutral
20607   PAT502  TP53    1.0   0.0     loss
20613   PAT502  TP53    7.0   2.0  neutral
21854   PAT532  TP53    1.0   0.0     loss
21857   PAT532  TP53    4.0  20.0     gain
23505   PAT571  TP53    1.0  16.0     gain
23514   PAT571  TP53   10.0   4.0  neutral
27391   PAT663  TP53    1.0   0.0     loss
27405   PAT663  TP53   15.0   2.0  neutral
29411   PAT712  TP53    1.0   0.0     loss
29413   PAT712  TP53    3.0   1.0  neutral
29933   PAT725  TP53    1.0   9.0     gain
29934   PAT725  TP53    2.0   3.0  neutral
35819   PAT867  TP53    1.0   9.0     gain
35820   PAT867  TP53    2.0   4.0  neutral
41857  PAT1012  TP53    1.0   0.0     loss
41863  PAT1012  TP53    7.0   1.0  neutral
43369  PAT1049  TP53    1.0   1.0  neutral
43370  PAT1049  TP53    2.0   0.0     loss
46559  PAT1126  TP53    1.0   0.0     loss
46562  PAT1126  TP53    4.0   3.0  neutral
47422  PAT1148  TP53    1.0   1.0  neutral
47424  PAT1148  TP53    3.0   0.0     loss
51598  PAT1247  TP53    1.0   3.0  neutral
51603  PAT1247  TP53    6.0   0.0     loss

In [15]:
#drops all duplicate neutral probes, shows only first probe and conflict
conflicts_EGFR = EGFR_call_search[EGFR_call_search.duplicated('orderid', keep = False)]
conflicts_EGFR

orderid  gene  probe   cnv     call
12        PAT0  EGFR    1.0   2.0  neutral
13        PAT0  EGFR    2.0   0.0     loss
3363     PAT81  EGFR    1.0   3.0  neutral
3367     PAT81  EGFR    5.0   0.0     loss
7335    PAT178  EGFR    1.0   3.0  neutral
7337    PAT178  EGFR    3.0  20.0     gain
7420    PAT180  EGFR    1.0   6.0  neutral
7424    PAT180  EGFR    5.0  20.0     gain
7928    PAT192  EGFR    1.0   0.0     loss
7940    PAT192  EGFR   13.0   1.0  neutral
8175    PAT198  EGFR    1.0   4.0  neutral
8192    PAT198  EGFR   18.0   8.0     gain
9117    PAT221  EGFR    1.0   6.0  neutral
9121    PAT221  EGFR    5.0  20.0     gain
12967   PAT315  EGFR    1.0   4.0  neutral
12971   PAT315  EGFR    5.0  20.0     gain
14249   PAT345  EGFR    1.0   8.0     gain
14250   PAT345  EGFR    2.0   2.0  neutral
17756   PAT432  EGFR    1.0   2.0  neutral
17758   PAT432  EGFR    3.0  20.0     gain
21361   PAT520  EGFR    1.0  13.0     gain
21366   PAT520  EGFR    6.0   7.0  neutral
21869   PAT532  EGFR    1.0   2.0  neutral
21874   PAT532  EGFR    6.0  20.0     gain
22291   PAT542  EGFR    1.0   5.0  neutral
22295   PAT542  EGFR    5.0  20.0     gain
33454   PAT810  EGFR    1.0   5.0  neutral
33461   PAT810  EGFR    8.0  20.0     gain
49501  PAT1198  EGFR    1.0   4.0  neutral
49505  PAT1198  EGFR    5.0  20.0     gain
51959  PAT1256  EGFR    1.0   3.0  neutral
51962  PAT1256  EGFR    4.0  20.0     gain
53220  PAT1289  EGFR    1.0   5.0  neutral
53238  PAT1289  EGFR   19.0  20.0     gain
58697  PAT1421  EGFR    1.0   6.0  neutral
58709  PAT1421  EGFR   13.0  20.0     gain
62387  PAT1510  EGFR    1.0   2.0  neutral
62389  PAT1510  EGFR    3.0  20.0     gain

In [16]:
conflicts_CDKN2A = CDKN2A_call_search[CDKN2A_call_search.duplicated('orderid', keep = False)]
conflicts_CDKN2A

orderid    gene  probe  cnv     call
7445    PAT180  CDKN2A    1.0  1.0  neutral
7446    PAT180  CDKN2A    2.0  0.0     loss
13548   PAT328  CDKN2A    1.0  1.0  neutral
13549   PAT328  CDKN2A    2.0  0.0     loss
16563   PAT403  CDKN2A    1.0  1.0  neutral
16564   PAT403  CDKN2A    2.0  0.0     loss
23251   PAT564  CDKN2A    1.0  2.0  neutral
23252   PAT564  CDKN2A    2.0  0.0     loss
32121   PAT777  CDKN2A    1.0  1.0  neutral
32123   PAT777  CDKN2A    3.0  0.0     loss
40950   PAT989  CDKN2A    1.0  0.0     loss
40951   PAT989  CDKN2A    2.0  1.0  neutral
51984  PAT1256  CDKN2A    1.0  1.0  neutral
51987  PAT1256  CDKN2A    4.0  0.0     loss
58722  PAT1421  CDKN2A    1.0  2.0  neutral
58723  PAT1421  CDKN2A    2.0  0.0     loss

In [17]:
conflicts_full = pd.concat([conflicts_TP53, conflicts_EGFR, conflicts_CDKN2A], axis =0)
conflicts_full

orderid    gene  probe   cnv     call
5229    PAT127    TP53    1.0   0.0     loss
5243    PAT127    TP53   15.0   2.0  neutral
12823   PAT312    TP53    1.0   4.0  neutral
12834   PAT312    TP53   12.0   0.0     loss
18835   PAT458    TP53    1.0   8.0     gain
18836   PAT458    TP53    2.0   3.0  neutral
20607   PAT502    TP53    1.0   0.0     loss
20613   PAT502    TP53    7.0   2.0  neutral
21854   PAT532    TP53    1.0   0.0     loss
21857   PAT532    TP53    4.0  20.0     gain
23505   PAT571    TP53    1.0  16.0     gain
23514   PAT571    TP53   10.0   4.0  neutral
27391   PAT663    TP53    1.0   0.0     loss
27405   PAT663    TP53   15.0   2.0  neutral
29411   PAT712    TP53    1.0   0.0     loss
29413   PAT712    TP53    3.0   1.0  neutral
29933   PAT725    TP53    1.0   9.0     gain
29934   PAT725    TP53    2.0   3.0  neutral
35819   PAT867    TP53    1.0   9.0     gain
35820   PAT867    TP53    2.0   4.0  neutral
41857  PAT1012    TP53    1.0   0.0     loss
41863  PAT1012    TP53    7.0   1.0  neutral
43369  PAT1049    TP53    1.0   1.0  neutral
43370  PAT1049    TP53    2.0   0.0     loss
46559  PAT1126    TP53    1.0   0.0     loss
46562  PAT1126    TP53    4.0   3.0  neutral
47422  PAT1148    TP53    1.0   1.0  neutral
47424  PAT1148    TP53    3.0   0.0     loss
51598  PAT1247    TP53    1.0   3.0  neutral
51603  PAT1247    TP53    6.0   0.0     loss
...        ...     ...    ...   ...      ...
22291   PAT542    EGFR    1.0   5.0  neutral
22295   PAT542    EGFR    5.0  20.0     gain
33454   PAT810    EGFR    1.0   5.0  neutral
33461   PAT810    EGFR    8.0  20.0     gain
49501  PAT1198    EGFR    1.0   4.0  neutral
49505  PAT1198    EGFR    5.0  20.0     gain
51959  PAT1256    EGFR    1.0   3.0  neutral
51962  PAT1256    EGFR    4.0  20.0     gain
53220  PAT1289    EGFR    1.0   5.0  neutral
53238  PAT1289    EGFR   19.0  20.0     gain
58697  PAT1421    EGFR    1.0   6.0  neutral
58709  PAT1421    EGFR   13.0  20.0     gain
62387  PAT1510    EGFR    1.0   2.0  neutral
62389  PAT1510    EGFR    3.0  20.0     gain
7445    PAT180  CDKN2A    1.0   1.0  neutral
7446    PAT180  CDKN2A    2.0   0.0     loss
13548   PAT328  CDKN2A    1.0   1.0  neutral
13549   PAT328  CDKN2A    2.0   0.0     loss
16563   PAT403  CDKN2A    1.0   1.0  neutral
16564   PAT403  CDKN2A    2.0   0.0     loss
23251   PAT564  CDKN2A    1.0   2.0  neutral
23252   PAT564  CDKN2A    2.0   0.0     loss
32121   PAT777  CDKN2A    1.0   1.0  neutral
32123   PAT777  CDKN2A    3.0   0.0     loss
40950   PAT989  CDKN2A    1.0   0.0     loss
40951   PAT989  CDKN2A    2.0   1.0  neutral
51984  PAT1256  CDKN2A    1.0   1.0  neutral
51987  PAT1256  CDKN2A    4.0   0.0     loss
58722  PAT1421  CDKN2A    1.0   2.0  neutral
58723  PAT1421  CDKN2A    2.0   0.0     loss

[84 rows x 5 columns]

To find how many cases had conflicts, we now just want unique patients and gene entries from  this table

In [18]:
conflicts_full.nunique()

orderid    38
gene        3
probe      14
cnv        13
call        3
dtype: int64

38 unique IDs tells us that some of these case had more than two conflicts

In [19]:
conflict_count = conflicts_full.groupby('orderid').orderid.count()
conflict_count.sort_values()

orderid
PAT0       2
PAT345     2
PAT403     2
PAT432     2
PAT458     2
PAT502     2
PAT520     2
PAT542     2
PAT564     2
PAT571     2
PAT663     2
PAT712     2
PAT725     2
PAT777     2
PAT81      2
PAT810     2
PAT328     2
PAT867     2
PAT315     2
PAT221     2
PAT1012    2
PAT1049    2
PAT1126    2
PAT1148    2
PAT1198    2
PAT1247    2
PAT312     2
PAT127     2
PAT1510    2
PAT178     2
PAT192     2
PAT198     2
PAT1289    2
PAT989     2
PAT532     4
PAT1256    4
PAT1421    4
PAT180     4
Name: orderid, dtype: int64

Patients 532, 1256, 1421, and 180 all had 4 values in the list, meaning they had two sets of conflicting calls each

In [20]:
PAT532_cases = conflicts_full.loc[conflicts_full['orderid'] == 'PAT532']
PAT1256_cases  = conflicts_full.loc[conflicts_full['orderid'] == 'PAT1256']
PAT1421_cases = conflicts_full.loc[conflicts_full['orderid'] == 'PAT1421']
PAT180_cases  = conflicts_full.loc[conflicts_full['orderid'] == 'PAT180']
multi_conflict_pat= pd.concat([PAT532_cases, PAT1256_cases, PAT1421_cases, PAT180_cases], axis =0)
multi_conflict_pat.sort_index(inplace = True)
multi_conflict_pat
#printunique values from orderid
multi_conflict_pat['orderid'].unique()


array(['PAT180', 'PAT532', 'PAT1256', 'PAT1421'], dtype=object)


#### Cases with multiple calls are considered conflicting. These conflicts can come in different combinations.
#### What is the most common type of conflict in the dataset?

In [21]:
df = conflicts_full[['orderid','call']]
df
   

orderid     call
5229    PAT127     loss
5243    PAT127  neutral
12823   PAT312  neutral
12834   PAT312     loss
18835   PAT458     gain
18836   PAT458  neutral
20607   PAT502     loss
20613   PAT502  neutral
21854   PAT532     loss
21857   PAT532     gain
23505   PAT571     gain
23514   PAT571  neutral
27391   PAT663     loss
27405   PAT663  neutral
29411   PAT712     loss
29413   PAT712  neutral
29933   PAT725     gain
29934   PAT725  neutral
35819   PAT867     gain
35820   PAT867  neutral
41857  PAT1012     loss
41863  PAT1012  neutral
43369  PAT1049  neutral
43370  PAT1049     loss
46559  PAT1126     loss
46562  PAT1126  neutral
47422  PAT1148  neutral
47424  PAT1148     loss
51598  PAT1247  neutral
51603  PAT1247     loss
...        ...      ...
22291   PAT542  neutral
22295   PAT542     gain
33454   PAT810  neutral
33461   PAT810     gain
49501  PAT1198  neutral
49505  PAT1198     gain
51959  PAT1256  neutral
51962  PAT1256     gain
53220  PAT1289  neutral
53238  PAT1289     gain
58697  PAT1421  neutral
58709  PAT1421     gain
62387  PAT1510  neutral
62389  PAT1510     gain
7445    PAT180  neutral
7446    PAT180     loss
13548   PAT328  neutral
13549   PAT328     loss
16563   PAT403  neutral
16564   PAT403     loss
23251   PAT564  neutral
23252   PAT564     loss
32121   PAT777  neutral
32123   PAT777     loss
40950   PAT989     loss
40951   PAT989  neutral
51984  PAT1256  neutral
51987  PAT1256     loss
58722  PAT1421  neutral
58723  PAT1421     loss

[84 rows x 2 columns]

In [22]:
df = df.groupby('orderid').agg({'call':','.join
    }).reset_index()
df

orderid                       call
0      PAT0               neutral,loss
1   PAT1012               loss,neutral
2   PAT1049               neutral,loss
3   PAT1126               loss,neutral
4   PAT1148               neutral,loss
5   PAT1198               neutral,gain
6   PAT1247               neutral,loss
7   PAT1256  neutral,gain,neutral,loss
8    PAT127               loss,neutral
9   PAT1289               neutral,gain
10  PAT1421  neutral,gain,neutral,loss
11  PAT1510               neutral,gain
12   PAT178               neutral,gain
13   PAT180  neutral,gain,neutral,loss
14   PAT192               loss,neutral
15   PAT198               neutral,gain
16   PAT221               neutral,gain
17   PAT312               neutral,loss
18   PAT315               neutral,gain
19   PAT328               neutral,loss
20   PAT345               gain,neutral
21   PAT403               neutral,loss
22   PAT432               neutral,gain
23   PAT458               gain,neutral
24   PAT502               loss,neutral
25   PAT520               gain,neutral
26   PAT532     loss,gain,neutral,gain
27   PAT542               neutral,gain
28   PAT564               neutral,loss
29   PAT571               gain,neutral
30   PAT663               loss,neutral
31   PAT712               loss,neutral
32   PAT725               gain,neutral
33   PAT777               neutral,loss
34    PAT81               neutral,loss
35   PAT810               neutral,gain
36   PAT867               gain,neutral
37   PAT989               loss,neutral

In [23]:
conflicts_split =  df['call'].str.split(',',3, expand=True)

conflicts_split

0        1        2     3
0   neutral     loss     None  None
1      loss  neutral     None  None
2   neutral     loss     None  None
3      loss  neutral     None  None
4   neutral     loss     None  None
5   neutral     gain     None  None
6   neutral     loss     None  None
7   neutral     gain  neutral  loss
8      loss  neutral     None  None
9   neutral     gain     None  None
10  neutral     gain  neutral  loss
11  neutral     gain     None  None
12  neutral     gain     None  None
13  neutral     gain  neutral  loss
14     loss  neutral     None  None
15  neutral     gain     None  None
16  neutral     gain     None  None
17  neutral     loss     None  None
18  neutral     gain     None  None
19  neutral     loss     None  None
20     gain  neutral     None  None
21  neutral     loss     None  None
22  neutral     gain     None  None
23     gain  neutral     None  None
24     loss  neutral     None  None
25     gain  neutral     None  None
26     loss     gain  neutral  gain
27  neutral     gain     None  None
28  neutral     loss     None  None
29     gain  neutral     None  None
30     loss  neutral     None  None
31     loss  neutral     None  None
32     gain  neutral     None  None
33  neutral     loss     None  None
34  neutral     loss     None  None
35  neutral     gain     None  None
36     gain  neutral     None  None
37     loss  neutral     None  None

In [24]:
#made two new columns firstcombo and secondcombo that are the two original pairs of columns concatenated
conflicts_split['firstcombo'] = conflicts_split[0] + conflicts_split[1]
conflicts_split['secondcombo'] = conflicts_split[2] + conflicts_split[3]
conflicts_clean = conflicts_split[['firstcombo','secondcombo']]

print(conflicts_clean.firstcombo.value_counts())
print(conflicts_clean.secondcombo.value_counts())

neutralgain    13
neutralloss    10
lossneutral     8
gainneutral     6
lossgain        1
Name: firstcombo, dtype: int64
neutralloss    3
neutralgain    1
Name: secondcombo, dtype: int64


In [25]:
#adding totals of each combo type from the two newly made columns
neutral_gain_combos = conflicts_clean.firstcombo.value_counts()['neutralgain']+conflicts_clean.firstcombo.value_counts()['gainneutral']\
+conflicts_clean.secondcombo.value_counts()['neutralgain']
neutral_loss_combos = conflicts_clean.firstcombo.value_counts()['lossneutral']+conflicts_clean.firstcombo.value_counts()['neutralloss']\
+conflicts_clean.secondcombo.value_counts()['neutralloss']
loss_gain_combos = conflicts_clean.firstcombo.value_counts()['lossgain']

print(neutral_gain_combos,neutral_loss_combos,loss_gain_combos)

20 21 1


Neutral-Loss conflicts are the most common

#### How can I turn the messiness of the cnvs and calls into a call that is useful?

In [26]:
df_TP53.describe()

probe           cnv
count  20214.000000  20214.000000
mean       7.307213      2.277728
std        4.088664      1.416745
min        1.000000      0.000000
25%        4.000000      2.000000
50%        7.000000      2.000000
75%       10.000000      2.000000
max       15.000000     20.000000

A meaningful call would probably involve giving a weight to the cnv.<br>
According to the instructions, a cnv of 1-7 is relatively neutral, and should be weighted as such<br>
CNV of 0 is a total absence of copies, and should carry the maximum negative weight<br>
CNV of 8-20(20 being the max found in all of the patients) would have ascending weight.

In [27]:
TP53_PAT1012 = df_TP53.loc[df_TP53.orderid == 'PAT1012'] 
TP53_PAT1012

orderid  gene  probe  cnv     call
41857  PAT1012  TP53    1.0  0.0     loss
41858  PAT1012  TP53    2.0  0.0     loss
41859  PAT1012  TP53    3.0  0.0     loss
41860  PAT1012  TP53    4.0  0.0     loss
41861  PAT1012  TP53    5.0  0.0     loss
41862  PAT1012  TP53    6.0  0.0     loss
41863  PAT1012  TP53    7.0  1.0  neutral
41864  PAT1012  TP53    8.0  1.0  neutral
41865  PAT1012  TP53    9.0  1.0  neutral
41866  PAT1012  TP53   10.0  1.0  neutral
41867  PAT1012  TP53   11.0  1.0  neutral
41868  PAT1012  TP53   12.0  1.0  neutral
41869  PAT1012  TP53   13.0  1.0  neutral
41870  PAT1012  TP53   14.0  1.0  neutral
41871  PAT1012  TP53   15.0  1.0  neutral

In [28]:
TP53_1 = df_TP53.loc[df_TP53.probe == 1] 
TP53_1.describe()

probe          cnv
count  1525.0  1525.000000
mean      1.0     2.253115
std       0.0     1.352624
min       1.0     0.000000
25%       1.0     2.000000
50%       1.0     2.000000
75%       1.0     2.000000
max       1.0    20.000000

In [29]:
TP53_PAT1012mod = TP53_PAT1012
TP53_PAT1012mod 

orderid  gene  probe  cnv     call
41857  PAT1012  TP53    1.0  0.0     loss
41858  PAT1012  TP53    2.0  0.0     loss
41859  PAT1012  TP53    3.0  0.0     loss
41860  PAT1012  TP53    4.0  0.0     loss
41861  PAT1012  TP53    5.0  0.0     loss
41862  PAT1012  TP53    6.0  0.0     loss
41863  PAT1012  TP53    7.0  1.0  neutral
41864  PAT1012  TP53    8.0  1.0  neutral
41865  PAT1012  TP53    9.0  1.0  neutral
41866  PAT1012  TP53   10.0  1.0  neutral
41867  PAT1012  TP53   11.0  1.0  neutral
41868  PAT1012  TP53   12.0  1.0  neutral
41869  PAT1012  TP53   13.0  1.0  neutral
41870  PAT1012  TP53   14.0  1.0  neutral
41871  PAT1012  TP53   15.0  1.0  neutral

Here I make a function with arbitrary weights that follow the logic stated earlier.<br>
The mean and evaluation are not made in this function because there may be some pattern in the location or frequency of the weighted values that could potentially offer an insight(purely hypothetical)

In [30]:

#Made up weights for each of the cnvs based on 0 being a total deletion, and the max based on the max across all patients,
#cnv of 1-7 is just weighted as 0(neutral)

def convert_to_weight(cnvs):
    cnvs['cnv'].replace([0], -1000,inplace=True)
    cnvs['cnv'].replace([1,2,3,4,5,6,7], 0, inplace=True)
    cnvs['cnv'].replace([8,9,10], 100,inplace=True)
    cnvs['cnv'].replace([11,12,13,14],300,inplace=True)
    cnvs['cnv'].replace([15,16,17,18,19,20], 1000,inplace=True)
    
    return cnvs     
    

In [31]:
convert_to_weight(TP53_PAT1012mod)

C:\Users\raobr\Anaconda3\lib\site-packages\pandas\core\generic.py:5886: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


orderid  gene  probe     cnv     call
41857  PAT1012  TP53    1.0 -1000.0     loss
41858  PAT1012  TP53    2.0 -1000.0     loss
41859  PAT1012  TP53    3.0 -1000.0     loss
41860  PAT1012  TP53    4.0 -1000.0     loss
41861  PAT1012  TP53    5.0 -1000.0     loss
41862  PAT1012  TP53    6.0 -1000.0     loss
41863  PAT1012  TP53    7.0     0.0  neutral
41864  PAT1012  TP53    8.0     0.0  neutral
41865  PAT1012  TP53    9.0     0.0  neutral
41866  PAT1012  TP53   10.0     0.0  neutral
41867  PAT1012  TP53   11.0     0.0  neutral
41868  PAT1012  TP53   12.0     0.0  neutral
41869  PAT1012  TP53   13.0     0.0  neutral
41870  PAT1012  TP53   14.0     0.0  neutral
41871  PAT1012  TP53   15.0     0.0  neutral

An arbitrary function that performs some calculations and returns an evaluation on the significant of the cnvs

In [32]:
def evaluate_weight(mean):
    x = mean/200
    print(x)
    if int(round(x)) == 0:
        return 'neutral'
    elif abs(int(round(x))) >=2:
        return 'significant'
    else:
        return 'potentially significant'
    


In [33]:
mean1 = TP53_PAT1012mod.cnv.mean() 
evaluate_weight(mean1)

-2.0


'significant'

#### Lets try a patient with a conflict involving gain


In [34]:
TP53_PAT458 = df_TP53.loc[df_TP53.orderid == 'PAT458'] 
TP53_PAT458

orderid  gene  probe  cnv     call
18835  PAT458  TP53    1.0  8.0     gain
18836  PAT458  TP53    2.0  3.0  neutral
18837  PAT458  TP53    3.0  3.0  neutral
18838  PAT458  TP53    4.0  3.0  neutral
18839  PAT458  TP53    5.0  3.0  neutral
18840  PAT458  TP53    6.0  3.0  neutral
18841  PAT458  TP53    7.0  3.0  neutral
18842  PAT458  TP53    8.0  3.0  neutral
18843  PAT458  TP53    9.0  3.0  neutral
18844  PAT458  TP53   10.0  3.0  neutral
18845  PAT458  TP53   11.0  3.0  neutral
18846  PAT458  TP53   12.0  3.0  neutral

In [35]:
TP53_PAT458mod = TP53_PAT458

In [36]:
convert_to_weight(TP53_PAT458mod)

C:\Users\raobr\Anaconda3\lib\site-packages\pandas\core\generic.py:5886: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)


orderid  gene  probe    cnv     call
18835  PAT458  TP53    1.0  100.0     gain
18836  PAT458  TP53    2.0    0.0  neutral
18837  PAT458  TP53    3.0    0.0  neutral
18838  PAT458  TP53    4.0    0.0  neutral
18839  PAT458  TP53    5.0    0.0  neutral
18840  PAT458  TP53    6.0    0.0  neutral
18841  PAT458  TP53    7.0    0.0  neutral
18842  PAT458  TP53    8.0    0.0  neutral
18843  PAT458  TP53    9.0    0.0  neutral
18844  PAT458  TP53   10.0    0.0  neutral
18845  PAT458  TP53   11.0    0.0  neutral
18846  PAT458  TP53   12.0    0.0  neutral

In [37]:
mean2 = TP53_PAT458mod.cnv.mean()
evaluate_weight(mean2)

0.04166666666666667


'neutral'

With these two patients as examples, this conversion of all of their calls to weights seems to sufficiently illustrate what a real, science-based weighting and function would aim to do.